sqlglot/optimizer/qualify_tables.py

```diff
@@ -114,5 +114,10 @@ def _qualify(table: exp.Table) -> None:
                 if isinstance(udtf, exp.Values) and not table_alias.columns:
                     for i, e in enumerate(udtf.expressions[0].expressions):
                         table_alias.append("columns", exp.to_identifier(f"_col_{i}"))
+            else:
+                for table in scope.cte_table_references[name]:
```

Comment:
why do you need cte table refs, why not find all unqualified tabkes?

Reply:
Can do that too, this way we simply avoid the traversal.

Reply:
This approach seemed like almost 0 cost and doesn't affect queries that don't have CTE Table refs. The other way around we'd have to traverse _all_ queries and check for unaliased tables which can be a significant overhead.

Reply:
ok we can keep it but let’s not use default dict

Reply:
since others will access it. default dict is ok for internal

Reply:
is there no way to get access to tables here?

Reply:
it's tricky to access the tables in `qualify_tables` because we loop over the sources and they're created in such a way that we store scopes for CTEs, so there's no reference to the actual table at hand.

> since others will access it. default dict is ok for internal

Makes sense, will update

Reply:
https://github.com/tobymao/sqlglot/pull/2865/commits/61e40cab9167a8944960a59209b5f98d8aaa0969